### ISSUE TITLE
Gradient of `numpy.linalg.eigh` with Hermitian matrix input

### ISSUE THREAD

**Issue by momchilmm** [HIPS/autograd#522]
It seems to me that `autograd.numpy.linalg.eigh` accepts a Hermitian matrix on the forward pass, but the gradient for that case is wrong, as it is only implemented for real symmetric matrices. I have a simple example of how this is a problem:
```
import autograd.numpy as npa
from autograd import grad

def fn(a):
    # Put `a` as an imaginary off-diagonal number of a Hermitian matrix
    mat = npa.array([[1, 1j*a], [-1j*a, 2]]) 
    eigs = npa.linalg.eigh(mat)[0] # diagonalize
    return eigs[0] # return the lowest eigenvalue

a = 2. # Some random test value

print('Autograd gradient:  ', grad(fn)(a))
print('Numerical gradient: ', (fn(a + 1e-5)-fn(a))/1e-5) 
```

Returns:
```
Autograd gradient:   0.0
Numerical gradient:  -0.9701426428199155
```
 
It's also written in [this line](https://github.com/HIPS/autograd/blob/387c373115ddd54cff2c8ba6a9fc619f28639cfb/autograd/numpy/linalg.py#L105) that a symmetric matrix is expected.

I don't really understand the code there so it's not clear to me if there's an easy extension, but some further insight would be greatly appreciated!

**Comment by momchilmm** [comment]
I just realized there's an [open pull request](https://github.com/HIPS/autograd/pull/462) fixing this. Would be great if it gets merged to avoid future issues!

**Comment by duvenaud** [comment]
The problem with that pull request is that it doesn't have any tests.  Although I suppose one could argue that the current code already silently fails for complex matrices, and if we accept this PR it at least have a chance of being right.  Hmmm.

**Comment by momchilmm** [comment]
I see. I looked into that but it seems to be a bigger can of worms than I thought. For example, I also don't think that the current tests for "lower" and "upper" eigh are correct, because the input matrix to the test is always symmetrized first. Instead if one inputs a non-symmetric (non-hermitian) matrix, the output of eigh depends only either on the lower part or only on the upper part, and the gradient w.r.t. the other part must correspondingly be zero.

I made a bit of progress but need to work on this more when I find more time.
